atlasdb-client/src/test/java/com/palantir/atlasdb/table/description/NameMetadataDescriptionTest.java

```diff
@@ -0,0 +1,164 @@
+import com.palantir.atlasdb.encoding.PtBytes;
+import com.palantir.atlasdb.protos.generated.TableMetadataPersistence;
+import com.palantir.atlasdb.ptobject.EncodingUtils;
+import com.palantir.common.exception.PalantirRuntimeException;
+
+public class NameMetadataDescriptionTest {
+    private static final NameMetadataDescription SIMPLE_NAME_METADATA_DESCRIPTION = NameMetadataDescription.safe(
+            "string", ValueType.STRING);
+    private static final NameMetadataDescription MULTIPART_NAME_METADATA_DESCRIPTION = NameMetadataDescription.create(
+            ImmutableList.of(
+                    new NameComponentDescription.Builder()
+                            .componentName("alpha")
+                            .type(ValueType.VAR_LONG)
+                            .byteOrder(TableMetadataPersistence.ValueByteOrder.DESCENDING)
+                            .build(),
+                    NameComponentDescription.safe("beta", ValueType.SIZED_BLOB),
+                    NameComponentDescription.safe("gamma", ValueType.VAR_STRING),
+                    NameComponentDescription.safe("omega", ValueType.STRING)));
+
+    private static final byte[] SAMPLE_ALPHA = EncodingUtils.flipAllBits(ValueType.VAR_LONG.convertFromJava(42L));
+    private static final byte[] SAMPLE_BETA = ValueType.SIZED_BLOB.convertFromJava(new byte[5]);
+    private static final byte[] SAMPLE_GAMMA = ValueType.VAR_STRING.convertFromString("boo");
+    private static final byte[] SAMPLE_OMEGA = ValueType.STRING.convertFromString("O(n)");
+    private static final byte[] SAMPLE_ROW = EncodingUtils.add(SAMPLE_ALPHA, SAMPLE_BETA, SAMPLE_GAMMA, SAMPLE_OMEGA);
+    private static final byte[] SAMPLE_ROW_PREFIX = EncodingUtils.add(SAMPLE_ALPHA, SAMPLE_BETA);
+
+    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+
+    @Test
+    public void parseAndRenderAreInverses_Simple() {
+        byte[] row = PtBytes.toBytes("theData");
+        assertThat(SIMPLE_NAME_METADATA_DESCRIPTION.parseFromJson(
+                SIMPLE_NAME_METADATA_DESCRIPTION.renderToJson(row), false)).containsExactly(row);
+    }
+
+    @Test
+    public void extraFieldsAreTolerated() {
+        String extraFieldJson = "{\"string\": \"tom\", \"extraneous\": \"another\"}";
+        byte[] result = SIMPLE_NAME_METADATA_DESCRIPTION.parseFromJson(extraFieldJson, false);
+        assertThat(result).containsExactly(PtBytes.toBytes("tom"));
+    }
+
+    @Test
+    // TODO (jkong): This looks ridiculous, but I don't wish to change existing behaviour.
+    public void duplicateFieldsAreTolerated() {
+        String invalidJson = "{\"string\": \"tom\", \"string\": \"robert\"}";
+        byte[] result = SIMPLE_NAME_METADATA_DESCRIPTION.parseFromJson(invalidJson, false);
+        assertThat(result).satisfiesAnyOf(
+                (bytes) -> assertThat(bytes).containsExactly(PtBytes.toBytes("tom")),
+                (bytes) -> assertThat(bytes).containsExactly(PtBytes.toBytes("robert")));
```

Comment:
I can't quite understand this test - it implies to me that you could get either result non-deterministically. Is this really the case currently?

Reply:
Currently you'll always get `robert` (in general, the last key-value mapping). This is dependent on the JSON library, though - while we could write just that assertion I don't think it's something we can guarantee.

I think my comment was probably a bit misleading in that "don't wish to change" referred to "tolerate duplicate fields", not specifically which key gets returned.

Reply:
I think the comments you have provided make this a lot clearer. Thanks!